File: python/2.7/strategy/Duck.py
from FlyBehavior import *
from QuackBehavior import *

class Duck(object):
	def __init__(self):
		self.flyBehavior = FlyBehavior
		self.quackBehavior = QuackBehavior

	def display(self):
		pass

	def performFly(self):
		self.flyBehavior.fly()

	def performQuack(self):
		self.quackBehavior.quack()

	def prformBehavior(self):
		self.quackBehavior.quack();

	def setFlyBehavior(self, fb):
		self.flyBehavior = fb

	def setQuackBehavior(self, qb):
		self.QuackBehavior = qb

	def swim(self):
		print "All ducks float, even decoys!"

class MallardDuck(Duck):
	def __init__(self):
		super(MallardDuck, self).__init__()
		self.flyBehavior = FlyWithWings()
		self.quackBehavior = Quack()

	def display(self):
		print "I'm a real Mallard duck"

class ModelDuck(Duck):
    def __init__(self):
        super(ModelDuck, self).__init__()
        self.flyBehavior = FlyNoWay()
        self.quackBehavior = Quack()

    def display(self):
        print "I'm ModelDuck"

File: js/strategy/Duck.js
var FlyWithWings = require('./FlyBehavior').FlyWithWings;
var FlyNoWay = require('./FlyBehavior').FlyNoWay;

var Quack = require('./QuackBehavior').Quack;

var Duck = function() {
    this.flyBehavior = null;
    this.quackBehavior = null;
}
Duck.prototype.display = function() {
    throw new Error("This method must be overwritten!\n");
}
Duck.prototype.performFly = function() {
    this.flyBehavior.fly();
}
Duck.prototype.performQuack = function() {
    this.quackBehavior.quack();
}
Duck.prototype.setFlyBehavior = function(fb) {
    this.flyBehavior = fb;
}
Duck.prototype.setQuackBehavior = function(qb) {
    this.quackBehavior = qb;
}
Duck.prototype.swim = function() {
    console.log("All ducks float, even decoys!\n");
}

var MallardDuck = function() {
    Duck.apply(this);
    this.flyBehavior = new FlyWithWings();
    this.quackBehavior = new Quack();
}
MallardDuck.prototype = new Duck();
MallardDuck.prototype.display = function() {
    console.log("I'm a real Mallard duck\n");
}

var ModelDuck = function() {
    Duck.apply(this);
    this.flyBehavior = new FlyNoWay();
    this.quackBehavior = new Quack();
}
ModelDuck.prototype = new Duck();
ModelDuck.prototype.display = function() {
    console.log("I'm a model duck\n");
}